src/stash/src/lib.rs

```diff
@@ -0,0 +1,465 @@
+            named_params! {"$upper": upper, "$arrangement_id": self.arrangement_id},
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Advances the since frontier to the specified value.
+    ///
+    /// The provided `since` must be greater than or equal to the current since
+    /// frontier but less than or equal to the current upper frontier.
+    ///
+    /// Intuitively, this method performs logical compaction. Existing entries
+    /// whose time is less than `since` are fast-forwarded to `since`.
+    pub fn compact(&self, since: i64) -> Result<(), StashError> {
+        let mut conn = self.conn.lock().expect("lock poisoned");
+        let tx = conn.transaction()?;
+        assert!(since <= self.upper_tx(&tx)?);
+        assert!(self.since_tx(&tx)? <= since);
+        tx.execute(
+            "UPDATE sinces SET since = $since WHERE arrangement_id = $arrangement_id",
+            named_params! {"$since": since, "$arrangement_id": self.arrangement_id},
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Consolidates entries less than the since frontier.
+    ///
+    /// Intuitively, this method performs physical compaction. Existing
+    /// key–value pairs whose time is less than the since frontier are
+    /// consolidated together when possible.
+    pub fn consolidate(&mut self) -> Result<(), StashError> {
+        let mut conn = self.conn.lock().expect("lock poisoned");
+        let tx = conn.transaction()?;
+        let since = self.since_tx(&tx)?;
+        tx.execute(
+            "INSERT INTO data (arrangement_id, key, value, time, diff)
+             SELECT arrangement_id, key, value, $since, sum(diff) FROM data
+             WHERE arrangement_id = $arrangement_id AND time < $since
+             GROUP BY key, value
+             ON CONFLICT (arrangement_id, key, value, time) DO UPDATE SET diff = diff + excluded.diff",
+            named_params! {
+                "$arrangement_id": self.arrangement_id,
+                "$since": since,
+            },
+        )?;
+        tx.execute(
+            "DELETE FROM data WHERE arrangement_id = $arrangement_id AND time < $since",
+            named_params! {
+                "$arrangement_id": self.arrangement_id,
+                "$since": since,
+            },
+        )?;
+        tx.commit()?;
+        Ok(())
+    }
+
+    /// Reports the current since frontier.
+    pub fn since(&self) -> Result<i64, StashError> {
+        let mut conn = self.conn.lock().expect("lock poisoned");
+        let tx = conn.transaction()?;
+        let since = self.since_tx(&tx)?;
+        tx.commit()?;
+        Ok(since)
+    }
+
+    /// Reports the current upper frontier.
+    pub fn upper(&self) -> Result<i64, StashError> {
+        let mut conn = self.conn.lock().expect("lock poisoned");
+        let tx = conn.transaction()?;
+        let upper = self.upper_tx(&tx)?;
+        tx.commit()?;
+        Ok(upper)
+    }
```

Comment:
If at all possible, it would be great if this returned a `Result<Option<i64>, StashError>` to capture the range of `Antichain<i64>` values, specifically the empty set. The empty set is what means "this collection is complete", and is something we end up using for e.g. non-tailed file reads. There is probably an expressivity regression if we lose track of this information, where we end up stuck with a fully read file that can't say "done" but instead say "I might say something at `i64::maximum()`.